### DISCUSSION TITLE
Cannot Flash Back to Stock Firmware - Dreamer NX

### DISCUSSION THREAD

**Discussion by Danosimo** [moonglow/FlashForge_Marlin#82] [Q&A] [answered]
Hi,
I have flashed the Marlin firmware in my dreamer NX and it works fine, but now I cannot flash back to stock firmware, looks like the PID/VID have changed (USB\VID_0483&PID_5740\20763370414E) my printer shows as FF407ZG CDC in FS mode when connected via USB. 
When I first tried to flash nx_2.0.9.2_la_11252021.bin,  FF Firmware writer could not finish the process (some error after coping file via USB) so I used the Flashprint firmware upgrade replacing stock file with Marlin .bin with same name.
It upgraded fine and now I have Marlin running, but cannot flash back to stock, the FF Firmware writer does not detect Dreamer and the Flashprint does not find it either. I have copied the original files to the internal flash but I cannot trigger the update. I have tried to use the RESET EPROM in marlin advanced menu but it beeps and nothing happens.
What can I do?. I appreciate any help here.
Daniel

**Comment by moonglow** [comment]
Hi @Danosimo  !

For future requests please provide me more info step by step with file names and so on, sometimes i can understand what user try to do ,but some times not, because some sensitive info was missed. First of all i highly recommends to use [latest FF writer ](https://github.com/moonglow/FlashForge_Marlin/wiki/Windows-instalation-guide) or `ff_flash_firmware.py` ( scripts works under Linux ) and Flashprint only as fallback solution. First of all you need to know that printer has a bootloader and you need to flash printer only at bootloader stage you can flash *any* firmware while bootloader runing. Marlin itself shows in system as virtual serial port and does not allow you to flash firmware while application ( Marlin ) running that's why you saw different VID/PID and so on. `EEPROM reset` in Marlin menu will just revert back to default Marlin settings. So you need to try update firmware while bootloader was running ( push start button in FF writer or start script and only when power-on your printer ). If your system is Windows you must also be sure what you install drivers for bootloader ( bootloader has different VID/PID, so when unknown devise was appear you need to be ready to install FF drivers for it ) Flashprint also will able to flash firmware while bootloader was running. Another fallback solution to use Update trigger from menu:

> I have copied the original files to the internal flash

Which name or path you use ? For `Firmware update trigger` you must put original firmware with name `dreamerx.bin` ( in case of Dreamer NX )  to `sys` folder of internal uSD card ( card located at printer motherboard and you need to remove bottom plate to get access it ). Next you need to push `Firmware update trigger` menu and reboot your printer.

**Comment by Danosimo** [comment]
Thank you for the quick reply Moonglow.
Yes, I was aware of the process. I did leave out some information, but it is just what you expected. I do not get an unknown device in Windows, it actually shows the Flashforge dreamer with correct VID/PID for a few seconds, then Marling loads and device changes.

I have copied the proper files in the SYS directory of the internal flash, I put the original firmware file with the name dreamerx.bin but I cannot find that option. How can I  "use Update trigger from menu"?, is it in the Marlin menu somewhere?. Is there a way to force the boot stage and NOT load the Marlin firmware? Like using the touch screen (like for the screen calibration)

 I have boot 1.4 installed and as you mentioned, when the Dreamer NX boots for a brief moment the proper PID&VID does appear in the system (windows 10 device manager) for a few seconds, but FF will not detect it (I see a brief pause after I hit START and boot the printer - dots display stop moving - then printer it goes into marling and FF times out. I am using ff_fwriter_08_10_2021.zip

 I am familiar with the process and I understand how the VID/PID changes when Marlin firmware is loaded. The thing is that FF does not detect the printer when booted even though I see the device (win device manager briefly while booting). I also have made sure the Flashforge drivers are installed, this worked fine when I flashed Marling firmware, however I used the Flashprint to flash the Marlin because FF complained after coping the flasd files at the end of the process (it said it could not set some flag I think).

Please let me know when I can find the upgrade firmware trigger option

**Comment by Danosimo** [comment]
One more thing, I flashed the latest 2.0.9.2 for dreamer NX and I am using original firmware renamed as dreamerx.bin in the /sys directorty of the internal SD card to revert back to stock, used ff_fwriter_08_10_2021.zip for the flashing and the Dreamer_1.9xxxx.bin original firmware from Flashforge

**Comment by moonglow** [comment]
>it actually shows the Flashforge dreamer with correct VID/PID for a few seconds, then Marling loads and device changes.

Yep, it is correct bootloader starts first and then it start Marlin firmware. You try to connect at bootloader stage via my tools or via FlashPrint itself ( before Marlin be loaded ).

>Is there a way to force the boot stage and NOT load the Marlin firmware

Yep, you can use my tools or Flashprint to interrupt booting ( just try to connect while device be in bootloader mode )

>How can I "use Update trigger from menu"?, is it in the Marlin menu somewhere?

Ah... sorry that's my mistake, trigger menu only enabled for Dreamer ( not NX ) and Dremel machines, because some of them has old bootloader what will not allow user to update via bootloader function.

>I have boot 1.4 installed and as you mentioned, when the Dreamer NX boots for a brief moment the proper PID&VID does appear in the system (windows 10 device manager) for a few seconds, but FF will not detect it

Hm.... strange can you tell me which VID/PID apear ? maybe they changed something.... ( you can try also to update via Flashprint at this stage )

**Comment by Danosimo** [comment]
So like I said, when I had the stock firmware, your tool (FF writer) did detect the printer booting and started the process. It would send the file and at the end of the process it seemed that it could not set the flag for firmware upgrade (the error was cannot set upgrade or something like that). After this I would reboot the machine and nothing happened. So I had to rename the Marlin file to trick Flashprint to use it with the same name as the original file (in the firmware directory of flashprint). Then Flashprint used it to upgrade, no problem and Marlin is working fine.

Could it be a timing issue and that is why FF writer cannot detect the printer while in boot mode briefly? It does show in the device manager with the proper PID/VID, why does it no detect? I am goint to test with a Linux machine (I have tested with 3 PC windows10), but 
The key problem here is that I cannot set the upgrade flag, because after setting that the .bin file placement is trivial. I mean, if I have no choice I can pull out my programmer and go directly for the eprom flash, but I have no backup of my original flash, not sure if the stock binary is enough, I would think so.

**Reply by moonglow** [reply]
>Then Flashprint used it to upgrade, no problem and Marlin is working fine.

You can try to update firmware via Flashprint from bootloader mode too, but better to fix my main tool :) 

>It does show in the device manager with the proper PID/VID, why does it no detect? 

Please provide me that VID/PID please

>I am goint to test with a Linux machine

Thats is my main setup, i think all will be OK ( but you can install Flashprint for Linux first to make sure what udev rules was updated to get access to USB without root rights ).

> I can pull out my programmer and go directly for the eprom flash, but I have no backup of my original flash

I think you no need to do that, because Linux way will works in most cases, but if you want to write flash via JTAG you need to decrypt firmware before with my [Firmware Tool](https://github.com/moonglow/flashforge_firmware_tool). Make backup first. Btw you will write firmware to MCU FLASH not EEPROM

**Comment by Danosimo** [comment]
During boot I get (in dev manager) - Flashforge Dreamer NX Printer - and hardware IDs - USB\VID_0315&PID_000A&REV_0200 - but no way to catch the upgrade to sync.
So I will try first in Linux cause as you said, it should work fine, all looks good , so no reason for it. If I cannot get it to work in Linux I will ask again.
By the way, did you test the BLTouch for leveling?
Thanks alot for all the help, great work on this. Let me know if I can help somehow
Daniel

**Reply by moonglow** [reply]
> During boot I get (in dev manager) - Flashforge Dreamer NX Printer - and hardware IDs - USB\VID_0315&PID_000A&REV_0200 - but no way to catch the upgrade to sync.

Really strange thing... i will try to find a windows machine to test it all

>  By the way, did you test the BLTouch for leveling?

I got my BLTouch ,but i did not find a time to test it all, we have many free pins to connect it, so i think in theory it will works

Any news ? did you revert back your firmware ?

**Comment by Danosimo** [comment]
I have not tried yet as I am enjoying the Marlin so far I have to confess. I am now more concern about updating Marlin in the future since I cannot catch the bootloader now. The strange thing is that when I first tried to flash Marlin with your tool from Flashforge original firmware, the FF writer did detect the printer after start and it would send the file to the printer. After the process, I assume when it is time to set the update flag, it would give me an error. So I used the Flashprint renaming the Marlin file to trick it and it worked. Now with the same windows machine, same cable and same everything it refuses to update with either (FF writer or Flashprint).
The only thing I see is that when FF writer is waiting for the printer, when Windows detects the bootloader (it shows dreamer device with proper PID/VID), the little dots under "wait for printer" in the FF writer tool, actually stop moving until the device manager changes the device to the STMFF407ZG CDC in FS mode (Serial Com port). Does that tell you anything?. 
Thanks again!

**Comment by Danosimo** [comment]
Hi, I tried to flash from a debian and it did start the loop after detecting the printer, but I get error on line 75 of the script, not sure what the problem is. It does stop the boot from loading Marling, but Cannot finish the process.
I tried to go back to flashprint in windows while bootloader was stopped but no luck
Any idea what to do? 
Thanks in advance

**Reply by moonglow** [reply]
>STMFF407ZG CDC in FS mode (Serial Com port). Does that tell you anything?.

This is standard Marlin USB serial port interface, you can open it and communicate with printer with many tools which was supported by Marlin.

> Any idea what to do?

Try to rerun script without rebooting the printer or try to restart printer and try again, very strange thing anyway, because i do it really frequency times ( because i test new build and so on ) other users use it too, but some has a problems with my Windows tool, but Linux way works.... If printer stops booting that mean what bootloader get in control and wait for other commands. I really want to fix this issue.... 

p.s: my main machine is Dreamer NX with Bootloader 1.4 so i think we have totally same machines, that's why it looks strange for me.

**Comment by Danosimo** [comment]
I am going to do a full install of debian instead of using a live version as I have, having to configure everything every time I reboot makes things harder.
I was wondering if just sending the trigger for flash on next boot would work for me since the .bin file is already copied to the internal SD.
Can this be done using an ADB type of tool? Or a GCODE command?. 
Anyway, I need to figure this out, I am still using the printer but eventually need to be able to flash it.
Let me know,
Daniel

**Comment by Danosimo** [comment, marked as answer]
Well, happy ending for me, I tried again a couple of times and it worked just fine. I am going to try to go back and forth a couple of times but it all looks good. Maybe the timing is a little sensitive?. Same PC so nothing changed, you were right,  just had to try a few more times I guess. Is the second control loop always necessary? Have you had success initializing control loop only once?.
Anyway, its a great thing that you give us the choice to use Marlin, so a big THANK YOU Moonglow!!!

**Reply by moonglow** [reply]
Huh that's great news ! You can try to plays with that script to make it works always and i will test it too maybe we can improve it. Only such tests with different envelopment can provide more stable product. Thanx ! I will try to fix it all soon
